### DISCUSSION TITLE
Custom serialization type safety

### DISCUSSION THREAD

**Discussion by noahw3** [mikro-orm/mikro-orm#4093] [Q&A] [answered]
This is very much related to https://github.com/mikro-orm/mikro-orm/discussions/1675.

At a high level, what I want to do is match the `toJSON` or `toObject` response type to what is actually produced at runtime.

I'm able to achieve what I want _at runtime_ using several different methods, such as setting a custom property serializer or using virtual properties. However, as far as I can tell none of these changes are actually reflected in the type - it's always just `EntityDTO<EntityName>`.

For a simple example, I have an entity definition with a `ManyToOne` relationship with another entity. It looks something like:
```
@Entity()
export class EntityName extends BaseEntity<EntityName, "id"> {
  @PrimaryKey({ type: "uuid", defaultRaw: "uuid_generate_v4()" })
  id: string;

  @ManyToOne(() => Tenant)
  @Index()
  tenant: Tenant;
}
```
This works fine for the most part. However when I want to serialize this entity to return from an API, I only want to include the `tenantId`. I can specify a custom serializer, like `{ serializer: tenant => tenant.id, serializedName: 'tenantId' }`. However the return type of `toJSON` doesn't change; `tenant` is still defined and `tenantId` is missing. My return type is not correctly satisfied, and the type definition allows me to access a nonexistent field. 

I can potentially fix this by overriding the `toJSON` function and adjusting the return type as necessary, or by adding my own custom wrapper. Ideally neither of these should be required though, and are prone to breaking if the entity definition changes.

I can't tell if I'm missing something obvious or if this is expected.

**Comment by B4nan** [comment]
You can't expect any information from decorator options to propagate to the type, this is purely runtime information. So yes, this is expected, custom serialization via decorator options can't have an effect on type-level results. 

Serialization respects populate hints, but for that you need to use the reference wrapper. That way, the information whether you loaded something (via populate hint) will be kept and respected in `EntityDTO` type. This is what the `Loaded` type is about.

<img width="537" alt="image" src="https://user-images.githubusercontent.com/615580/222860062-888ec0f4-c059-4064-9674-346ac9c44b3f.png">
<img width="527" alt="image" src="https://user-images.githubusercontent.com/615580/222860207-b473c353-2726-4346-9069-502e031a36a1.png">

(the screenshots are from v6, but it should work the same in v5)

https://mikro-orm.io/docs/type-safe-relations

**Reply by noahw3** [reply]
Thanks for the confirmation on how decorators work.

I understand that populate hints can help load the joined entities (and ensure correct typing), but I don't see how it can be used to actually change the response format. In my example above, I could specify a populate hint of `['tenant.id']`, but that doesn't help me map it to `tenantId` in my response.

It seems that the best solution in this case is to just write my own serialization function?

**Comment by noahw3** [comment, marked as answer]
Following up, I was largely able to get this to work with a combination of `hidden: true`, a getter with `persist: false`, and using `OptionalProps`. So the example above looks like:

```
@Entity()
export class EntityName extends BaseEntity<EntityName, "id"> {
  [OptionalProps]?: "tenantId";
  
  @PrimaryKey({ type: "uuid", defaultRaw: "uuid_generate_v4()" })
  id!: string;

  @ManyToOne({
    entity: () => Tenant,
    hidden: true,
    ref: true,
  })
  tenant: Ref<Tenant>;
  
  @Property({ persist: false })
  get tenantId() {
    return this.tenant.id;
  }
}
```

`OptionalProps` is necessary to avoid having to pass a `tenantId` in create calls. The one thing that this doesn't resolve is that `toJSON` is still typed such that `tenant` will be defined even though it won't be serialized, so you have to be careful with that.

**Reply by B4nan** [reply]
Maybe we should also add `HiddenProps` symbol? Given v6 has strict partial loading support, this sounds like a good companion.

**Reply by noahw3** [reply]
As in, being able to explicitly mark properties that won't be included in the serialized type? If so, then I like it.

**Reply by B4nan** [reply]
https://github.com/mikro-orm/mikro-orm/commit/aa4b3b7c59e045b9f1e3ec8f61b0fef728c02670
